app/models/user.rb

```
@@ -0,0 +1,12 @@
+class User < ApplicationRecord
+  has_many :highlights, dependent: :destroy
+  has_many :user_sources, dependent: :destroy
+
+  def increment_num_highlights(by: 1)
+    update(num_highlights: [num_highlights + by, 0].max)
+  end
+
+  def increment_num_annotation_characters(by: 1)
+    update(num_annotation_characters: [num_annotation_characters + by, 0].max)
```

interested in hearing how increments into negative territory would happen; and rationale for this approach vs validating that the number stays non negative.

---

it does make sense to fail early if this happens, perhaps thru a validation.  i just figured that the room for error here was higher and we'd not worry too much about it.  If you feel that we need to tighten it up..perhaps you're right and we should have a validation here.  what would you like to do?

---

reimplemented as a rails.logger.error before_validation.